Disable VB.NET templates for .NET Core

rodrmoya opened the pull request.

rodrmoya commented:
NOTE: should find a better condition to disable VB.NET support I guess?

mhutch commented:
I wonder if this should done in a more generic way, e.g. `<Condition id="FeatureSwitch" name="VBNetDotnetCoreTemplates" />` and a `MD_FEATURES_ENABLED="Feature1;Feature2"` env var.

mhutch commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.FeatureConfiguration/GlobalFeatures.cs:
```diff
@@ -0,0 +1,52 @@
+	{
+		public static class Languages
+		{
+			public static FeatureSwitch VisualBasicSupport = FeatureSwitchService.RegisterFeature (
```
This is *very* misleadingly named, as it doesn't control VB.NET support at all, it just controls the .NET Core templates

rodrmoya replied:
right, indeed

mhutch commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.FeatureConfiguration/GlobalFeatures.cs:
```diff
@@ -0,0 +1,52 @@
+		public override bool Evaluate (NodeElement conditionNode)
+		{
+			return Environment.GetEnvironmentVariables ().Contains ("MD_ENABLE_VISUALBASIC_SUPPORT")
+				|| BrandingService.ApplicationName.Contains ("MonoDevelop");
```
I don't think it's necessary to keep these templates enabled for MD. Also, the semantics of this are weird - I think the intention is "disable for VSMac" but it reads as "disable for any rebranded instance".

Therzok replied:
Maybe we can implement this extension point as returning false in Xamarin.Ide.

rodrmoya commented:
Changed it to be more generic and to explicitly disable it in md-addins -> https://github.com/xamarin/md-addins/pull/3988

mhutch commented on main/src/addins/MonoDevelop.DotNetCore/Properties/MonoDevelop.DotNetCore.addin.xml:
```diff
@@ -95,18 +95,20 @@
-			wizard="MonoDevelop.DotNetCore.ProjectTemplateWizard"
-			condition="UseNetCore21=true"
-			category="netcore/app/general"/>
+		<Condition id="FeatureSwitchCondition" name="VBNetDotnetCoreTemplates">
```
IMO "Condition" is redundant in `FeatureSwitchCondition`, it could just be `FeatureSwitch`

mhutch commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.FeatureConfiguration/GlobalFeatures.cs:
```diff
@@ -0,0 +1,53 @@
+				return true;
+			}
+
+			if (Environment.GetEnvironmentVariable ("MD_FEATURES_ENABLED")
```
Wonder whether it's worth caching the split string in a field

rodrmoya replied:
for normal situation (i.e. the IDE) yes, but for tests we set/unset the env vars for testing different situations, so I guess we should be splitting it always

mhutch commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.FeatureConfiguration/IFeatureSwitchCheck.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace MonoDevelop.Core.FeatureConfiguration
+{
+	public interface IFeatureSwitchCheck
```
Not a huge fan of this name, it seems to be defined from the point of view of the consumer, and doesn't make the purpose super obvious to me. `FeatureSwitchEnabler`, maybe?

mhutch commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.FeatureConfiguration/IFeatureSwitchCheck.cs:
```diff
@@ -0,0 +1,34 @@
+{
+	public interface IFeatureSwitchCheck
+	{
+		bool IsFeatureEnabled (string featureName);
```
I wonder whether this should return a `bool?` so that these can both enable _and_ disable features.

mhutch commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.FeatureConfiguration/GlobalFeatures.cs:
```diff
@@ -0,0 +1,53 @@
+
+			// Fallback to ask extensions, enabling by default
+			return !(AddinManager.GetExtensionObjects<IFeatureSwitchCheck> ("MonoDevelop.Core.FeatureConfiguration.IFeatureSwitchCheck")?
+				.Any (x => !x.IsFeatureEnabled (featureName)) ?? true);
```
It might make sense to make the default value (true/false) a per-feature thing. This would allow features to be on by default, and the extension could disable them

rodrmoya replied:
Well, that's what it does, right? It only disables the feature if one of the IFeatureSwitchCheck implementations returns false for that feature.

rodrmoya commented:
I'm not too happy with this implementation indeed, as I was trying to unify what we use for disabling/enabling Build Output (https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutput.cs#L47) with this, but because of the `NodeElement` argument in `ConditionType`'s `Evaluate`, can't think of a good way of unifying it. Maybe just make everything use the `FeatureSwitch` as BuildOutput does, and, on top of that, write the `ConditionType`-based class?

Or maybe use the `FeatureSwitchService` to register/query each individual feature toggle?

Commit:
[Core] Simplify FeatureSwitchService

Commit:
[DotNetCore] Enable VB.NET templates conditionally

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/750568

Commit:
[Ide] Adapt to FeatureSwitchService changes

Commit:
[Core] Allow passing a default value to FeatureSwitchService

So that it can be used if no explicit checkers are found. Also, add tests
for the new FeatureSwitchService

Commit:
[Tests] Fix BuildOutput feature toggle tests

rodrmoya commented:
I've started over and simplified the FeatureSwitchService implementation to use the IFeatureSwitchEnabler extension, so that all code is now unified and much simpler

mhutch commented:
Couple thoughts - since feature switches can now be enabled by default, maybe `IFeatureSwitchController` would be better? Also, having `IFeatureSwitchController.IsFeatureEnabled` return a `bool?` where null means "don't care" would be cleaner than passing in the default value as a parameter.

Commit:
[Core] Simplify FeatureSwitchService API

Return a bool? so that controllers can respond whether they want to
explicitly enable/disable that feature, or they just don't care about it.

Also, added several more specific tests.

rodrmoya commented:
@mhutch done! :)

rodrmoya commented:
@monojenkins backport release-7.8

rodrmoya commented:
@monojenkins backport release-7.7

rodrmoya commented:
Bug was marked for 7.7, but we don't have the VB.NET templates there, so no need to backport to 7.7
